### PR TITLE
Feat/shopify

### PR DESCRIPTION
**Pull request by Vladyslav317**

### CODE REVIEW

File: biome.jsonc [hunk 1]
```diff
@@ -1,91 +1,91 @@
 {
-  "$schema": "https://biomejs.dev/schemas/1.7.3/schema.json",
```
Comment (IgorrPopov) [inline biome.jsonc:2]: @balelkin already edited this file if I'm not mistaken

File: docker-compose.yml [hunk 2]
```diff
@@ -23,7 +23,7 @@ services:
       - redis
 
   postgres:
-    container_name: postgres
+    container_name: postgres-shopify-starter
```
Comment (IgorrPopov) [inline docker-compose.yml:26]: if you want to use unique container names make them consistent like with the app "shopify-starter-app"

File: package.json [hunk 3]
```diff
@@ -43,31 +43,39 @@
...
     "@types/joi": "^17.2.3",
     "@ucast/sql": "^1.0.0-alpha.1",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.1",
+    "cookie-parser": "^1.4.6",
     "express-basic-auth": "^1.2.1",
     "helmet": "^7.1.0",
+    "install": "^0.13.0",
     "ioredis": "^5.4.1",
+    "joi": "^17.13.1",
```
Comment (IgorrPopov) [inline package.json:58]: we already removed Join pull the main branch

File: package.json [hunk 4]
```diff
@@ -43,31 +43,39 @@
...
     "nodemailer-sendgrid": "^1.0.3",
+    "npm": "^10.8.1",
```
Comment (IgorrPopov) [inline package.json:60]: why do we need this? and what is it?

File: package.json [hunk 5]
```diff
@@ -43,31 +43,39 @@
...
     "reflect-metadata": "^0.2.0",
     "rxjs": "^7.8.1",
     "socket.io": "^4.7.5",
+    "uuid": "^10.0.0",
```
Comment (IgorrPopov) [inline package.json:64]: nodeJS has an internal function for UUID generation no need to use a third-party package

File: package.json [hunk 6]
```diff
@@ -43,31 +43,39 @@
...
     "@nestjs/cli": "^10.0.0",
     "@nestjs/schematics": "^10.0.0",
     "@nestjs/testing": "^10.0.0",
+    "@types/cookie-parser": "^1.4.7",
     "@types/express": "^4.17.17",
     "@types/jest": "^29.5.2",
     "@types/node": "^20.3.1",
     "@types/nodemailer": "^6.4.15",
     "@types/supertest": "^6.0.0",
+    "@types/uuid": "^9.0.8",
```
Comment (IgorrPopov) [inline package.json:78]: same

File: prisma/migrations/20240617121546_webhook/migration.sql [hunk 7]
```diff
@@ -0,0 +1,22 @@
+-- CreateEnum
+CREATE TYPE "EcommercePlatformsEnum" AS ENUM ('SHOPIFY', 'WOO_COMMERCE');
+
+-- CreateTable
+CREATE TABLE "webhooks" (
+    "id" UUID NOT NULL,
+    "webhookId" TEXT NOT NULL,
+    "ecommercePlatform" "EcommercePlatformsEnum" NOT NULL,
```
Comment (IgorrPopov) [inline prisma/migrations/20240617121546_webhook/migration.sql:8]: what's the point to have this?

File: prisma/schema.prisma [hunk 8]
```diff
@@ -28,7 +28,27 @@ model User {
   @@map("users")
 }
 
+model Webhook {
+  id                String                 @id(map: "pk_Webhook") @db.Uuid
+  webhookId         String
+  ecommercePlatform EcommercePlatformsEnum
```
Comment (IgorrPopov) [inline prisma/schema.prisma:34]: same

File: prisma/schema.prisma [hunk 9]
```diff
@@ -28,7 +28,27 @@ model User {
...
+  @@index([expiredAt])
+  @@map("webhooks")
+}
+
 enum RolesEnum {
   CUSTOMER
   SUPER_ADMIN
 }
+
+enum EcommercePlatformsEnum {
```
Comment (IgorrPopov) [inline prisma/schema.prisma:51]: no need to use it

File: src/config/shopify.config.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  readonly maxTries: number;
+  readonly maxPaginationLimit: number;
+  readonly appPurchaseOneTimeMinPrice: number;
+};
+
+export default registerAs(
+  'shopify',
+  (): ShopifyConfig => ({
+    apiKey: process.env.SHOPIFY_API_KEY,
+    apiSecret: process.env.SHOPIFY_API_SECRET,
```
Comment (IgorrPopov) [inline src/config/shopify.config.ts:17]: if add new ENVs where is .env.example file???

File: src/modules/app/app.module.ts [hunk 11]
```diff
@@ -4,12 +4,16 @@ import { mailerConfig } from '@config/mailer.config';
 import { ngrokConfig } from '@config/ngrok.config';
 import { RedisConfig, redisConfig } from '@config/redis.config';
 import { sentryConfig } from '@config/sentry.config';
+import shopifyConfig from '@config/shopify.config';
```
Comment (IgorrPopov) [inline src/modules/app/app.module.ts:7]: do not use default export/import

File: src/modules/email/email.service.ts [hunk 12]
```diff
@@ -68,6 +68,55 @@ export class EmailService {
...
+      'Customer redact data request',
+    );
+  }
+
+  private async sendCustomerData(
+    data: string,
+    pathName: string,
+    subject: string,
+  ): Promise<void> {
+    const appConfig = this.configService.get('app');
```
Comment (IgorrPopov) [inline src/modules/email/email.service.ts:92]: add type and generic type

File: src/modules/email/email.service.ts
Comment (IgorrPopov) [inline src/modules/email/email.service.ts:1]: why it's like this? const nodemailerSendgrid = require('nodemailer-sendgrid');

File: src/modules/email/email.service.ts [hunk 13]
```diff
@@ -68,6 +68,55 @@ export class EmailService {
...
+      });
+
+      await this.mailTransport.sendMail({
+        to: appConfig.adminEmail,
+        from: appConfig.emailFrom,
+        subject,
+        html: renderedHtml,
+      });
+    } catch (e) {
+      Logger.error(e);
```
Comment (IgorrPopov) [inline src/modules/email/email.service.ts:116]: like this **this.logger**

File: src/modules/email/templates/customer-data-request.html [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  <meta http-equiv="X-UA-Compatible" content="IE=edge">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <title>Shopify Customer Data Request</title>
+</head>
+
+<body style="font-family: Arial, sans-serif;">
+
+  <div style="max-width: 600px; margin: 0 auto; padding: 20px; background-color: #f4f4f4; border-radius: 10px; text-align: center;">
+
+    <img src="{{ logo }}" alt="Upboost Logo" width="200" height="60" style="display: block; margin: 0 auto 20px;">
```
Comment (IgorrPopov) [inline src/modules/email/templates/customer-data-request.html:15]: !!!!!!!!!!!

File: src/modules/email/templates/customer-redact.html [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+  <meta http-equiv="X-UA-Compatible" content="IE=edge">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <title>Shopify Customer Redact</title>
+</head>
+
+<body style="font-family: Arial, sans-serif;">
+
+  <div style="max-width: 600px; margin: 0 auto; padding: 20px; background-color: #f4f4f4; border-radius: 10px; text-align: center;">
+
+    <img src="{{ logo }}" alt="Upboost Logo" width="200" height="60" style="display: block; margin: 0 auto 20px;">
```
Comment (IgorrPopov) [inline src/modules/email/templates/customer-redact.html:15]: !!!!!!!!!!!!

File: src/modules/mandatory-webhook/mandatory-webhook.controller.ts [hunk 16]
```diff
@@ -0,0 +1,35 @@
+import { Controller, Post, RawBodyRequest, Req } from '@nestjs/common';
+import { Request } from 'express';
+import { ApiTags } from '@nestjs/swagger';
+import { MandatoryWebhookService } from './mandatory-webhook.service';
```
Comment (IgorrPopov) [inline src/modules/mandatory-webhook/mandatory-webhook.controller.ts:4]: with @

File: src/modules/mandatory-webhook/mandatory-webhook.controller.ts [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+@ApiTags('Mandatory-webhook')
```
Comment (IgorrPopov) [inline src/modules/mandatory-webhook/mandatory-webhook.controller.ts:6]: Mandatory Webhook

File: src/modules/shopify-app-install/shopify-app-install.controller.ts [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+  constructor(
+    private readonly shopifyAppInstallService: ShopifyAppInstallService,
+    private readonly configService: ConfigService,
+  ) {}
+
+  @Get('/install')
+  public async installApp(
+    @Query('shop') shop: string,
+    @Cookies()
+    { upBoostUserId: userId, upBoostWebShopId: webShopId }: CookiesType,
```
Comment (IgorrPopov) [inline src/modules/shopify-app-install/shopify-app-install.controller.ts:21]: !!!!!!!!!!

File: src/modules/shopify-app-install/shopify-app-install.controller.ts [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+    @Res() res: Response,
+  ): Promise<void> {
+    this.shopifyAppInstallService.validateHmac(req.query);
+
+    const { session } = req.query;
+    const { clientHost } = this.configService.get<AppConfig>('app');
+
+    if (shop && session) {
+      Logger.log(
+        `The app has been open for the shop: ${shop}. The request will be redirected to the UpBoost host: ${clientHost}`,
```
Comment (IgorrPopov) [inline src/modules/shopify-app-install/shopify-app-install.controller.ts:32]: !!!!!!!!!!!

File: src/modules/shopify-app-install/shopify-app-install.controller.ts [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+    await this.shopifyAppInstallService.beginAuth(req, res);
+  }
+
+  // TODO Callback should be added later when webshop logic is ready
+
+  // @Get('/callback')
+  // async callback(
+  //   @Query('shop') shop: string,
+  //   @Cookies()
+  //   { upBoostUserId: userId, upBoostWebShopId: webShopId }: CookiesType,
```
Comment (IgorrPopov) [inline src/modules/shopify-app-install/shopify-app-install.controller.ts:53]: !!!!!!!!!!!!!

File: src/modules/mandatory-webhook/mandatory-webhook.module.ts [hunk 21]
```diff
@@ -0,0 +1,14 @@
+import { Module } from '@nestjs/common';
+import { MandatoryWebhookService } from './mandatory-webhook.service';
```
Comment (IgorrPopov) [inline src/modules/mandatory-webhook/mandatory-webhook.module.ts:2]: import

File: src/modules/mandatory-webhook/mandatory-webhook.service.ts [hunk 22]
```diff
@@ -0,0 +1,181 @@
+import { EmailService } from '@modules/email/email.service';
+import { ShopifyAppInstallService } from '@modules/shopify-app-install/shopify-app-install.service';
+import { WebhookService } from '@modules/webhook/webhook.service';
+import { Injectable, Logger, RawBodyRequest, UnauthorizedException } from '@nestjs/common';
+import { EcommercePlatformsEnum, Webhook } from '@prisma/client';
+import { Request } from 'express';
+import { v4 as uuidv4 } from 'uuid';
```
Comment (IgorrPopov) [inline src/modules/mandatory-webhook/mandatory-webhook.service.ts:7]: no need to use uuid

File: src/modules/mandatory-webhook/mandatory-webhook.service.ts [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+          body: req.body,
+          headers: req.headers,
+        },
+      )}`,
+    );
+
+    try {
+      await this.uninstallApp(shopId);
+
+      Logger.debug(
```
Comment (IgorrPopov) [inline src/modules/mandatory-webhook/mandatory-webhook.service.ts:80]: like in the upper comment

File: src/modules/mandatory-webhook/mandatory-webhook.service.ts [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+      },
+      null,
+      2,
+    );
+
+    if (await this.webhookService.isDuplicate(webhookId)) return;
+
+    const { id: shopId } = req.body;
+
+    Logger.debug(
```
Comment (IgorrPopov) [inline src/modules/mandatory-webhook/mandatory-webhook.service.ts:118]: same

File: src/modules/mandatory-webhook/mandatory-webhook.service.ts [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+        )}`,
+      );
+    }
+  }
+
+  private saveWebhook(req: RawBodyRequest<Request>): Promise<Webhook> {
+    const expiredAt = new Date();
+
+    // Add 6 months
+    expiredAt.setMonth(expiredAt.getMonth() + 6);
```
Comment (IgorrPopov) [inline src/modules/mandatory-webhook/mandatory-webhook.service.ts:145]: with Postgres we don't have TTL

File: src/modules/mandatory-webhook/mandatory-webhook.service.ts [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+
+    return this.webhookService.create({
+      id: uuidv4(),
+      webhookId: req.headers['x-shopify-webhook-id']?.toString(),
+      body: req.body,
+      headers: req.headers,
+      topic: req.headers['x-shopify-topic']?.toString(),
+      ecommercePlatform: EcommercePlatformsEnum.SHOPIFY,
```
Comment (IgorrPopov) [inline src/modules/mandatory-webhook/mandatory-webhook.service.ts:153]: no need

File: src/modules/shopify-app-install/shopify-app-install.controller.ts [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+
+    if (shop && session) {
+      Logger.log(
+        `The app has been open for the shop: ${shop}. The request will be redirected to the UpBoost host: ${clientHost}`,
+      );
+
+      return res.redirect(`https://${clientHost}`);
+    }
+
+    Logger.debug(
```
Comment (IgorrPopov) [inline src/modules/shopify-app-install/shopify-app-install.controller.ts:38]: same

File: src/modules/shopify-app-install/shopify-app-install.module.ts [hunk 28]
```diff
@@ -0,0 +1,11 @@
+import { Module } from '@nestjs/common';
+import { ShopifyAppInstallController } from './shopify-app-install.controller';
```
Comment (IgorrPopov) [inline src/modules/shopify-app-install/shopify-app-install.module.ts:2]: same

File: src/modules/shopify-app-install/shopify-app-install.repository.ts [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+  shopifyApi,
+  ApiVersion,
+  Shopify,
+  BeginParams,
+} from '@shopify/shopify-api';
+import { Request, Response } from 'express';
+import { restResources } from '@shopify/shopify-api/rest/admin/2023-07';
+import { ConfigService } from '@nestjs/config';
+import { ShopifyConfig } from '@config/shopify.config';
+
```
Comment (IgorrPopov) [inline src/modules/shopify-app-install/shopify-app-install.repository.ts:13]: too many any in this file

File: src/modules/shopify-app-install/shopify-app-install.service.ts [hunk 30]
```diff
@@ -0,0 +1,57 @@
+import { Injectable, RawBodyRequest, UnauthorizedException } from '@nestjs/common';
+import { createHmac } from 'node:crypto';
+import { Request, Response } from 'express';
+import { ShopifyAppInstallRepository } from './shopify-app-install.repository';
```
Comment (IgorrPopov) [inline src/modules/shopify-app-install/shopify-app-install.service.ts:4]: same

File: src/modules/webhook/webhook.service.ts [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    return webhook;
+  }
+
+  public async isDuplicate(webhookId: string): Promise<boolean> {
+    const webhook: Webhook | null = await this.getOneByWebhookId(webhookId);
+
+    if (webhook) {
+      Logger.debug(
```
Comment (IgorrPopov) [inline src/modules/webhook/webhook.service.ts:27]: logger like in the other comments

File: tsconfig.json [hunk 32]
```diff
@@ -1,13 +1,13 @@
 {
   "compilerOptions": {
-    "module": "commonjs",
+    "moduleResolution": "nodenext",
```
Comment (IgorrPopov) [inline tsconfig.json:3]: explain me why?

File: biome.jsonc (hunk 1 above)
Comment (Vladyslav317) [reply biome.jsonc:2]: I helped him edit this file, but I tested it myself first and left the changes in my branch.

File: src/modules/email/email.service.ts
Comment (Vladyslav317) [reply src/modules/email/email.service.ts:1]: @balelkin Should it only work through require?

File: src/modules/shopify-app-install/shopify-app-install.repository.ts (hunk 29 above)
Comment (Vladyslav317) [reply src/modules/shopify-app-install/shopify-app-install.repository.ts:13]: Changed some but for session there will be type when we'll be ready to add webshops.

File: tsconfig.json (hunk 32 above)
Comment (Vladyslav317) [reply tsconfig.json:3]: commonjs - usually uses require/module.exports but nodenext - import, exports. More so, I needed it in order to import `import { restResources } from '@shopify/shopify-api/rest/admin/2023-07';` from shopify installation

File: src/modules/mandatory-webhook/mandatory-webhook.service.ts (hunk 25 above)
Comment (Vladyslav317) [reply src/modules/mandatory-webhook/mandatory-webhook.service.ts:145]: I know but if we decide to get all hooks that are already expired and work with them then I suppose this solution is ok. Or should I delete this field?

File: package.json [hunk 33]
```diff
@@ -43,14 +43,16 @@
...
     "@socket.io/redis-adapter": "^8.3.0",
     "@songkeys/nestjs-redis": "^10.0.0",
-    "@types/joi": "^17.2.3",
     "@ucast/sql": "^1.0.0-alpha.1",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.1",
+    "cookie-parser": "^1.4.6",
     "express-basic-auth": "^1.2.1",
     "helmet": "^7.1.0",
+    "install": "^0.13.0",
```
Comment (IgorrPopov) [inline package.json:55]: what is this???

File: prisma/schema.prisma [hunk 34]
```diff
@@ -28,6 +28,20 @@ model User {
   @@map("users")
 }
 
+model Webhook {
+  id                String                 @id(map: "pk_Webhook") @db.Uuid
+  webhookId         String
+  topic             String
+  body              Json
+  headers           Json
+  expiredAt         DateTime
```
Comment (IgorrPopov) [inline prisma/schema.prisma:37]: no need to have this

File: src/modules/mandatory-webhook/mandatory-webhook.service.ts [hunk 35]
```diff
@@ -0,0 +1,181 @@
+import { EmailService } from '@modules/email/email.service';
+import { ShopifyAppInstallService } from '@modules/shopify-app-install/shopify-app-install.service';
+import { WebhookService } from '@modules/webhook/webhook.service';
+import { Injectable, Logger, RawBodyRequest, UnauthorizedException } from '@nestjs/common';
+import { Webhook } from '@prisma/client';
+import { Request } from 'express';
+import { randomUUID } from 'node:crypto';
```
Comment (IgorrPopov) [inline src/modules/mandatory-webhook/mandatory-webhook.service.ts:7]: nodejs internal modules always go to the top of the file

File: src/modules/mandatory-webhook/mandatory-webhook.service.ts (hunk 25 above)
Comment (IgorrPopov) [reply src/modules/mandatory-webhook/mandatory-webhook.service.ts:147]: remove it

File: src/modules/shopify-app-install/shopify-app-install.controller.ts [hunk 36]
```diff
@@ -0,0 +1,109 @@
+import { Controller, Get, Logger, Query, Req, Res } from "@nestjs/common";
+import { Request, Response } from 'express';
+import { ApiTags } from '@nestjs/swagger';
+import { ShopifyAppInstallService } from "@modules/shopify-app-install/shopify-app-install.service";
+import { ConfigService } from "@nestjs/config";
+import { AppConfig } from "@config/app.config";
+import { Cookies, CookiesType } from "src/decorators/cookies.decorator";
```
Comment (IgorrPopov) [inline src/modules/shopify-app-install/shopify-app-install.controller.ts:7]: !!!!!!!!

File: src/modules/shopify-app-install/shopify-app-install.controller.ts (hunk 27 above)
Comment (IgorrPopov) [reply src/modules/shopify-app-install/shopify-app-install.controller.ts:38]: logger like in the upper files

File: src/modules/shopify-app-install/shopify-app-install.service.ts [hunk 37]
```diff
@@ -0,0 +1,58 @@
...
```
Comment (IgorrPopov) [inline src/modules/shopify-app-install/shopify-app-install.service.ts:4]: imports!!!!!!!1

**Pull request merged**